prov/usnic/src/usdf_cq.c

```diff
@@ -149,12 +149,106 @@ usdf_cq_readerr_soft(struct fid_cq *fcq, struct fi_cq_err_entry *entry,
 	return 1;
 }
 
+static inline ssize_t
+usdf_cq_sread_common(struct fid_cq *fcq, void *buf, size_t count, const void *cond,
+			int timeout, enum fi_cq_format format)
+{
+	struct usdf_cq *cq;
+	uint8_t *entry;
+	size_t entry_len;
+	struct fi_cq_entry *ctx_entry;
+	struct fi_cq_msg_entry *msg_entry;
+	struct fi_cq_data_entry *data_entry;
+	ssize_t ret;
+	size_t time_spent;
+
+	cq = cq_ftou(fcq);
+	if (cq->cq_comp.uc_status != 0) {
+		return -FI_EAVAIL;
+	}
+
+	switch (format) {
+	case FI_CQ_FORMAT_CONTEXT:
+		entry_len = sizeof(struct fi_cq_entry);
+		break;
+	case FI_CQ_FORMAT_MSG:
+		entry_len = sizeof(struct fi_cq_msg_entry);
+		break;
+	case FI_CQ_FORMAT_DATA:
+		entry_len = sizeof(struct fi_cq_data_entry);
+		break;
+	default:
+		return 0;
+	}
+
+	ret = 0;
+	entry = buf;
+
+	while (1) {
+		ret = usd_poll_cq(cq->c.hard.cq_cq, &cq->cq_comp);
+		if (ret == -EAGAIN) {
+			if (entry > (uint8_t *)buf)
+				break;
+			if (timeout > 0 && ((time_spent - timeout) >= 0))
+				break;
+			usleep(1000 * SREAD_SLEEP_TIME);
+			time_spent += SREAD_SLEEP_TIME;
+			continue;
+		}
+		if (cq->cq_comp.uc_status != 0) {
+			return -FI_EAVAIL;
+		}
+		switch (format) {
```

Comment:
this `switch` block can probably be made into a common `static inline` routine that is shared with `usdf_cq_read_common`